lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Delegation.hs

```diff
@@ -0,0 +1,464 @@
+-- "concurrent" states tracking stake keys, where the first one is identical to
+-- legacy single-stake key wallets.
+--
+-- Maybe that would help simplify `applyTx` and `setPortfolioOf`...
+data State
+    -- | No active stake keys. The initial state of a new wallet.
+    = Zero
+
+    -- | The first stake-key (index 0) is registered and either delegating or
+    -- about to be delegating.
+    | One
+
+    -- | There is more than one active stake keys. Can only be reached using
+    -- wallets with support for multiple stake keys.
+    | More
+        !(Index 'Soft 'AddressK)
+          -- nextKeyIx - the ix of the next unused key
+        PointerUTxO
+          -- ^ pointer utxo that need to be spent when changing state.
+        !Bool
+          -- ^ Is key 0 still registered? For compatibility with
+          -- single-stake-key wallets, we need to track this.
+          --
+          -- See the implementaiton of @applyTx@ for how it is used.
+    deriving (Eq, Show, Generic)
+
+instance NFData State
+
+instance (NFData (k 'AccountK XPub), NFData (k 'AddressK XPub))
+    => NFData (DelegationState k)
+
+deriving instance
+    ( Show (k 'AccountK XPub)
+    , Show (k 'AddressK XPub)
+    ) => Show (DelegationState k)
+
+deriving instance
+    ( Eq (k 'AccountK XPub)
+    , Eq (k 'AddressK XPub)
+    ) => Eq (DelegationState k)
+
+pointer :: DelegationState k -> Maybe PointerUTxO
+pointer (DelegationState _ (More _ p _)) = Just p
+pointer _ = Nothing
+
+keyAtIx
+    :: (SoftDerivation k)
+    => DelegationState k
+    -> Index 'Soft 'AddressK
+    -> k 'AddressK XPub
+keyAtIx s = deriveAddressPublicKey (rewardAccountKey s) MutableAccount
+
+nextKeyIx
+    :: DelegationState k
+    -> Index 'Soft 'AddressK
+nextKeyIx s = case state s of
+    Zero -> minBound
+    One -> succ minBound
+    More ix _ _ -> ix
+
+lastActiveIx
+    :: DelegationState k
+    -> Maybe (Index 'Soft 'AddressK)
+lastActiveIx s
+    | nextKeyIx s == minBound = Nothing
+    | otherwise               = Just $ pred $ nextKeyIx s
+
+data PointerUTxO = PointerUTxO { pTxIn :: TxIn, pCoin :: Coin }
+    deriving (Generic, Eq, Show)
+    deriving anyclass NFData
+
+-- | Returns the index corresponding to the payment key the `PointerUTxO`
+-- should be locked with for a portfolio of a given size @n@.
+--
+-- In our current implementation we require the `PointerUTxO` to be created in
+-- the @[0] -> [0,1] transition@, i.e. @nextKeyIx 1 -> nextKeyIx 2@.
+pointerIx
+    :: Int
+    -> Maybe (Index 'Soft 'AddressK)
+pointerIx 0 = Nothing
+pointerIx 1 = Nothing
+pointerIx n = Just $ toEnum n
+
+--------------------------------------------------------------------------------
+-- Constructors
+--------------------------------------------------------------------------------
+
+-- | Construct the initial delegation state.
+initialDelegationState
+    :: k 'AccountK XPub
+    -> DelegationState k
+initialDelegationState accK = DelegationState accK Zero
```

Comment:
Minor remark: this is quite far from the `DelegationState` definition. Since this is pretty much a smart constructor for that data-type I'd expect to see it close to the definition. (module exports also suggests that).